src/qutip_qip/circuit/circuitsimulator.py

```diff
@@ -599,41 +498,61 @@ def _evolve_state(self, U):
         U: Qobj
             unitary to be applied.
         """
-
+        if operation.name == "GLOBALPHASE":
+            # This is just a complex number.
+            U = np.exp(1.0j * operation.arg_value)
+        else:
+            # We need to use the circuit because the custom gates
+            # are still saved in circuit instance.
+            # This should be changed once that is deprecated.
+            U = self.qc._get_gate_unitary(operation)
+            U = expand_operator(
+                U,
+                dims=self.dims,
+                targets=operation.get_all_qubits(),
+            )
         if self.mode == "state_vector_simulator":
-            self._evolve_ket(U)
+            state = U * state
         elif self.mode == "density_matrix_simulator":
-            self._evolve_dm(U)
+            state = U * state * U.dag()
         else:
             raise NotImplementedError(
                 "mode {} is not available.".format(self.mode)
             )
+        return state
+
+    def _evolve_state_einsum(self, gate, state):
+        if gate.name == "GLOBALPHASE":
+            # This is just a complex number.
+            return np.exp(1.0j * gate.arg_value) * state
+        # Prepare the state tensor.
+        targets_indices = gate.get_all_qubits()
+        if isinstance(state, Qobj):
+            # If it is a Qobj, transform it to the array representation.
+            state = state.full()
+            # Transform the gate and state array to the corresponding
+            # tensor form.
+            state = state.reshape(self._tensor_dims)
+        # Prepare the gate tensor.
+        gate = self.qc._get_gate_unitary(gate)
+        gate_array = gate.full().reshape(gate.dims[0] + gate.dims[1])
+        # Compute the tensor indices and call einsum.
+        num_site = len(state.shape)
+        ancillary_indices = range(num_site, num_site + len(targets_indices))
+        index_list = range(num_site)
+        new_index_list = list(index_list).copy()
+        for j, k in enumerate(targets_indices):
+            new_index_list[k] = j + num_site
+        state = np.einsum(
+            gate_array,
```

Comment:
Question for the future -- could one in theory provide multiple gates at once here? Would that potentially be better / faster? Possibly with or without use of `np.einsum_path` to select an optimal contraction path?

Reply:
Yes, that could be faster if those gates are mostly local. Then it is not that easy to decide unless one first builds the directed graph of the circuit, because gates can be commuting. So instead of this, we should rather try to get tensor network to work and let it does the heavy lifting.